## Analyzing Spreadsheet Data with Python

Spreadsheets are one of the most common tools for storing and working with tabular data. However, when the data grow large, navigating row and columns manually becomes inefficient and error-prone. Python provides powerful libraries, such as Pandas, that make it easier to load, manipulate, and analyze spreadsheet data programmatically.

In this project, I demonstrate how to:
- Load spreadsheet data (.xlsx or .xls files) into Python.
- Explore and manipulate datasets using Pandas.
- Perform basic analysis such as subsetting, filtering, and calculating statistics (mean, median, maximum, etc.).
- Write processed data back into a spreadsheet file.

### Import the required packages

We start by importing the necessary python packages:
- `os` for interacting with the os (e.g., file path)
- `pandas` for data manipulation and analysis

In [1]:
import os
import pandas as pd

### 1. Load the data 

Next, we load the facilities data from the Excel file into a Pandas `DataFrame`.  
The dataset is stored in `airport_data.xlsx`, and we specifically read the **"Facilities"** sheet.  

In [2]:
df_facilities = pd.read_excel("./airport_data.xlsx", sheet_name="Facilities")

### 2. Exploring Data

Once the data is loaded, the next step is to inspect the dataset. By displaying the DataFrame, we can quickly verify that the data is imported correctly and gain an initial sense of its structure:

In [3]:
df_facilities

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
0      50009.*A        AIRPORT       'ADK     3/30/2017    AAL           NONE   
1     50016.1*A        AIRPORT       'AKK     3/30/2017    AAL           NONE   
2      50017.*A        AIRPORT       'Z13     3/30/2017    AAL           NONE   
3     50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL           NONE   
4      50020.*A        AIRPORT       'AKI     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL           NONE   
747    50920.*A        AIRPORT       'YAK     3/30/2017    AAL           NONE   
748   50925.1*A        AIRPORT       'A77     3/30/2017    AAL           NONE   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   
750    50990.*A        AIRPORT       'L20     3/30/2017    AAL           NONE   

    State StateName             County CountyState  ... AirportPositionSource  \
0      AK    ALASKA     ALEUTIANS WEST          AK  ...      3RD PARTY SURVEY   
1      AK    ALASKA      KODIAK ISLAND          AK  ...                  NACO   
2      AK    ALASKA             BETHEL          AK  ...                 STATE   
3      AK    ALASKA             BETHEL          AK  ...                   NaN   
4      AK    ALASKA             BETHEL          AK  ...                 STATE   
..    ...       ...                ...         ...  ...                   ...   
746    AK    ALASKA    SKAGWAY-YAKUTAT          AK  ...               FAA-EST   
747    AK    ALASKA    SKAGWAY-YAKUTAT          AK  ...      3RD PARTY SURVEY   
748    AK    ALASKA          KUSKOKWIM          AK  ...                   NaN   
749    AK    ALASKA  KETCHIKAN GATEWAY          AK  ...                   NaN   
750    AK    ALASKA        UPPER YUKON          AK  ...               FAA-EST   

    AirportPositionSourceDate AirportElevationSource  \
0                    00:00:00       3RD PARTY SURVEY   
1                    00:00:00                   NACO   
2                    00:00:00                  STATE   
3                         NaN                    NaN   
4                    00:00:00                  STATE   
..                        ...                    ...   
746                  00:00:00                    NaN   
747                  00:00:00       3RD PARTY SURVEY   
748                       NaN                    NaN   
749                       NaN                    NaN   
750                  00:00:00                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
0                     00:00:00                   NaN              HGR   
1                     00:00:00                   NaN              NaN   
2                     00:00:00                   NaN              NaN   
3                          NaN                   NaN              NaN   
4                     00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
746                        NaN                   NaN              TIE   
747                   00:00:00                   NaN              HGR   
748                        NaN                   NaN              NaN   
749                        NaN                   NaN              NaN   
750                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
0           CARGO             Y           PADK           SS-SR  
1             NaN             Y           PAKH             NaN  
2           CARGO           Y-L            NaN         SEE RMK  
3             NaN             N            NaN             NaN  
4           CARGO           Y-L           PFAK           SS-SR  
..            ...           ...            ...             ...  
746         CARGO           NaN            NaN        

The preview shows us the first and the last five rows of the dataset, including columns such as **`SiteNumber`**, **`Type`**, **`StateName`**, etc. 
This helps confirm that the spreadsheet was read correctly and gives us starting point for further analysis. 

#### 2.1. Selecting
electing data from a DataFrame is straightforward. We can access one or more columns by specifying their names in square brackets, enclosed in quotes.

Selecting specific columns enables us to examine a specific attribute of the dataset in more detail. Here, we'll retrieve the `Type` column, which contains information about the types of aviation facilities. 

In [4]:
df_facilities["Type"]

0            AIRPORT
1            AIRPORT
2            AIRPORT
3      SEAPLANE BASE
4            AIRPORT
           ...      
746    SEAPLANE BASE
747          AIRPORT
748          AIRPORT
749    SEAPLANE BASE
750          AIRPORT
Name: Type, Length: 751, dtype: object

Exploring individual columns helps us understand the types of data we are working with.  
In this case, examining the `"Type"` column gives insight into the categorical values of aviation facilities, which will guide filtering, grouping, or aggregation steps later.  
We can apply the Pandas `unique()` method on the series to quickly retrieve all distinct values present in the column.

In [5]:
df_facilities["Type"].unique()

array(['AIRPORT', 'SEAPLANE BASE', 'HELIPORT'], dtype=object)

The preview reveals that the "Type" column has three distinct values.
Understanding these categories is important because it informs how we might filter, group, or summarize the data in later analysis steps.
For example, knowing the types helps us analyze patterns or differences between facility categories.

#### 2.2. Filtering
After selecting a column and examining its values, we can filter the data to focus on specific entries.
For example, we will filter the "Type" column to include only rows corresponding to "SEAPLANE BASE".

In [6]:
[df_facilities["Type"] == "SEAPLANE BASE"] # return boolean values

[0      False
 1      False
 2      False
 3       True
 4      False
        ...  
 746     True
 747    False
 748    False
 749     True
 750    False
 Name: Type, Length: 751, dtype: bool]

In [7]:
df_facilities[df_facilities["Type"]=="SEAPLANE BASE"] # return real data where the condition is true

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
3     50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL           NONE   
5      50022.*C  SEAPLANE BASE       'KQA     3/30/2017    AAL           NONE   
8    50027.52*C  SEAPLANE BASE       'Z33     3/30/2017    AAL           NONE   
15    50031.5*C  SEAPLANE BASE      'AK81     3/30/2017    AAL           NONE   
24    50035.2*C  SEAPLANE BASE       '3C3     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
725  50877.32*C  SEAPLANE BASE       'MFN     3/30/2017    AAL           NONE   
734  50877.01*C  SEAPLANE BASE       '2X2     3/30/2017    AAL           NONE   
737    50905.*C  SEAPLANE BASE       '68A     3/30/2017    AAL           NONE   
746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL           NONE   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   

    State StateName               County CountyState  ...  \
3      AK    ALASKA               BETHEL          AK  ...   
5      AK    ALASKA       ALEUTIANS EAST          AK  ...   
8      AK    ALASKA          BRISTOL BAY          AK  ...   
15     AK    ALASKA        KODIAK ISLAND          AK  ...   
24     AK    ALASKA            ANCHORAGE          AK  ...   
..    ...       ...                  ...         ...  ...   
725    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
734    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
737    AK    ALASKA  WRANGELL-PETERSBURG          AK  ...   
746    AK    ALASKA      SKAGWAY-YAKUTAT          AK  ...   
749    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
3                     NaN                       NaN                    NaN   
5                   STATE                  00:00:00                  STATE   
8                     NaN                       NaN                    NaN   
15                    NaN                       NaN                    NaN   
24                  OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
725                 OWNER                  00:00:00                  OWNER   
734               FAA-EST                  00:00:00                  STATE   
737                   NaN                       NaN                    NaN   
746               FAA-EST                  00:00:00                    NaN   
749                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
3                          NaN                   NaN              NaN   
5                     00:00:00                   NaN              NaN   
8                          NaN                   NaN              TIE   
15                         NaN                   NaN              NaN   
24                    00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
725                   00:00:00                   NaN              NaN   
734                   00:00:00                   NaN              TIE   
737                        NaN                   NaN              NaN   
746                        NaN                   NaN              TIE   
749                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
3             NaN             N            NaN             NaN  
5             NaN             Y            NaN             NaN  
8             NaN           NaN            NaN             NaN  
15            NaN             N            NaN             NaN  
24            NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
725           NaN           Na

#### 2.3. Sorting
Sorting is the process of data rearrangement based on specific criteria. We often need so sort the data to better understand specific data in our dataset. In the next example we will sort a column name "ARPElevation" descendingly to see which the highest aviation. We will apply Pandas' `sort_values()` method.

In [8]:
df_facilities.sort_values(by="ARPElevation", ascending=False)

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
584   50739.1*A        AIRPORT      'AA09     3/30/2017    AAL           NONE   
630  50738.20*A        AIRPORT      '93AK     3/30/2017    AAL           NONE   
265   50324.6*A        AIRPORT       '4Z5     3/30/2017    AAL           NONE   
582  50738.01*A        AIRPORT      '2AK7     3/30/2017    AAL           NONE   
353  50439.18*A        AIRPORT      'AK80     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   
261    50321.*C  SEAPLANE BASE       'OOH     3/30/2017    AAL           NONE   
289  50385.06*H       HELIPORT      '73AK     3/30/2017    AAL           NONE   
714   50873.3*C  SEAPLANE BASE       'KWP     3/30/2017    AAL           NONE   
713    50872.*C  SEAPLANE BASE       'KWF     3/30/2017    AAL           NONE   

    State StateName               County CountyState  ...  \
584    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
630    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
265    AK    ALASKA  SOUTHEAST FAIRBANKS          AK  ...   
582    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
353    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
..    ...       ...                  ...         ...  ...   
749    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   
261    AK    ALASKA      SKAGWAY-YAKUTAT          AK  ...   
289    AK    ALASKA     ALEUTIAN ISLANDS          AK  ...   
714    AK    ALASKA        KODIAK ISLAND          AK  ...   
713    AK    ALASKA      PRINCE OF WALES          AK  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
584                   NaN                       NaN                    NaN   
630                 OWNER                  00:00:00                  OWNER   
265            FAA OE/AAA                  00:00:00             FAA OE/AAA   
582                   NaN                       NaN                    NaN   
353                 OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
749                   NaN                       NaN                    NaN   
261                   NaN                       NaN                    NaN   
289                 OWNER                  00:00:00                  OWNER   
714                   NaN                       NaN                    NaN   
713                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
584                        NaN                   NaN              NaN   
630                   00:00:00                   NaN              NaN   
265                   00:00:00                   NaN              NaN   
582                        NaN                   NaN              NaN   
353                   00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
749                        NaN                   NaN              NaN   
261                        NaN                   NaN              TIE   
289                   00:00:00                   NaN              NaN   
714                        NaN                   NaN              NaN   
713                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
584           NaN             Y            NaN             NaN  
630         CARGO             Y            NaN             NaN  
265           NaN             Y            NaN             NaN  
582           NaN           NaN            NaN             NaN  
353           NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
749         CARGO             

The preview above shows us the dataframe in descending order based on the elevation of the aviations. The are many questions come to mind after we the sorting is done, in this case, I want to demonstrate how to extract the first five aviation with the highest elevation.

In [9]:
df_facilities.sort_values(by="ARPElevation", ascending=False).head(n=5)

SiteNumber     Type LocationID EffectiveDate Region DistrictOffice State  \
584   50739.1*A  AIRPORT      'AA09     3/30/2017    AAL           NONE    AK   
630  50738.20*A  AIRPORT      '93AK     3/30/2017    AAL           NONE    AK   
265   50324.6*A  AIRPORT       '4Z5     3/30/2017    AAL           NONE    AK   
582  50738.01*A  AIRPORT      '2AK7     3/30/2017    AAL           NONE    AK   
353  50439.18*A  AIRPORT      'AK80     3/30/2017    AAL           NONE    AK   

    StateName               County CountyState  ... AirportPositionSource  \
584    ALASKA    MATANUSKA-SUSITNA          AK  ...                   NaN   
630    ALASKA    MATANUSKA-SUSITNA          AK  ...                 OWNER   
265    ALASKA  SOUTHEAST FAIRBANKS          AK  ...            FAA OE/AAA   
582    ALASKA    MATANUSKA-SUSITNA          AK  ...                   NaN   
353    ALASKA    MATANUSKA-SUSITNA          AK  ...                 OWNER   

    AirportPositionSourceDate AirportElevationSource  \
584                       NaN                    NaN   
630                  00:00:00                  OWNER   
265                  00:00:00             FAA OE/AAA   
582                       NaN                    NaN   
353                  00:00:00                  OWNER   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
584                        NaN                   NaN              NaN   
630                   00:00:00                   NaN              NaN   
265                   00:00:00                   NaN              NaN   
582                        NaN                   NaN              NaN   
353                   00:00:00                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
584           NaN             Y            NaN             NaN  
630         CARGO             Y            NaN             NaN  
265           NaN             Y            NaN             NaN  
582           NaN           NaN            NaN             NaN  
353           NaN             Y            NaN             NaN  

[5 rows x 103 columns]

We can adjust the number of rows returned by adjusting the number of n. 

Next, I'll demonstrate how to calculate the average of elevation by using `mean()` method.

In [10]:
df_facilities["ARPElevation"].mean()

np.float64(425.93209054593876)

The result tells us that the average elevation of the aviation is 425.9 feet. Next, we can use `max()` and `min()` function too find the highest and the lowest elevation value.

In [11]:
df_facilities["ARPElevation"].max()

np.int64(3984)

In [12]:
df_facilities["ARPElevation"].min()

np.int64(0)

As we can see from the above result that the highest elevation value is 3984 feet and the lowest is 0 feet.
 

### 3. Exporting processed file into excel.
After we finished processing the data and get the result we want, we often want to save the data for future use. In this case, we will export the data in a file called `seaplane.xlsx`.

 We first need to store the data in a variable with specific name that describe what the data is about.

In [13]:
df_seaplane = df_facilities[df_facilities["Type"] == "SEAPLANE BASE"]

Then, to save the data in excel, we can use `to_excel` method as follow:

In [14]:
df_seaplane.to_excel("./seaplane.xlsx")

In [15]:
df_seaplane

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
3     50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL           NONE   
5      50022.*C  SEAPLANE BASE       'KQA     3/30/2017    AAL           NONE   
8    50027.52*C  SEAPLANE BASE       'Z33     3/30/2017    AAL           NONE   
15    50031.5*C  SEAPLANE BASE      'AK81     3/30/2017    AAL           NONE   
24    50035.2*C  SEAPLANE BASE       '3C3     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
725  50877.32*C  SEAPLANE BASE       'MFN     3/30/2017    AAL           NONE   
734  50877.01*C  SEAPLANE BASE       '2X2     3/30/2017    AAL           NONE   
737    50905.*C  SEAPLANE BASE       '68A     3/30/2017    AAL           NONE   
746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL           NONE   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   

    State StateName               County CountyState  ...  \
3      AK    ALASKA               BETHEL          AK  ...   
5      AK    ALASKA       ALEUTIANS EAST          AK  ...   
8      AK    ALASKA          BRISTOL BAY          AK  ...   
15     AK    ALASKA        KODIAK ISLAND          AK  ...   
24     AK    ALASKA            ANCHORAGE          AK  ...   
..    ...       ...                  ...         ...  ...   
725    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
734    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
737    AK    ALASKA  WRANGELL-PETERSBURG          AK  ...   
746    AK    ALASKA      SKAGWAY-YAKUTAT          AK  ...   
749    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
3                     NaN                       NaN                    NaN   
5                   STATE                  00:00:00                  STATE   
8                     NaN                       NaN                    NaN   
15                    NaN                       NaN                    NaN   
24                  OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
725                 OWNER                  00:00:00                  OWNER   
734               FAA-EST                  00:00:00                  STATE   
737                   NaN                       NaN                    NaN   
746               FAA-EST                  00:00:00                    NaN   
749                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
3                          NaN                   NaN              NaN   
5                     00:00:00                   NaN              NaN   
8                          NaN                   NaN              TIE   
15                         NaN                   NaN              NaN   
24                    00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
725                   00:00:00                   NaN              NaN   
734                   00:00:00                   NaN              TIE   
737                        NaN                   NaN              NaN   
746                        NaN                   NaN              TIE   
749                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
3             NaN             N            NaN             NaN  
5             NaN             Y            NaN             NaN  
8             NaN           NaN            NaN             NaN  
15            NaN             N            NaN             NaN  
24            NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
725           NaN           Na